p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+ * times.  The PRE is made available to the Ingress programmable block
+ * using the same mechanism as packet_in. A corresponding Buffering
+ * and Queueing Engine (BQE) extern is defined for the Egress
+ * pipeline (@see BufferingQueueingEngine).
+ *
+ * Note: some of these operations may not be implemented as primitive
+ * operations on a certain target. However, All of the operations can
+ * be implemented as a combination of other operations. Applications
+ * that rely on non-primitive operations may incur significant
+ * performance penalties, however, they should be functionally
+ * correct.
+ *
+ * Semantics of behavior for multiple calls to PRE APIs
+ *
+ * The semantics of calling the PRE APIs is equivalent to setting
+ * intrinsic metadata fields/bits and assumming that the PRE looks up
+ * the fields in the following order: drop, truncate, multicast,
+ * clone, output_port.
+ *
+ * Following this semantics, examples of the behaviors are:
+ *
+ * - any call to drop in the pipeline will cause the packet, and all
+ *   potential clone copies (see below) to drop.
+ *
+ * - any call to truncate, will cause the packet (and its clones
+ *   \todo: check) to be truncated.
+ *
+ * - multiple calls to send_to_port -- the last call in the ingress
+ *   pipeline sets the output port.
+ *
+ * - multiple calls to multicast -- the last in the ingress pipeline
+ *   sets the multicast group
+ *
+ * - interleaving send_to_port and multicast -- the semantics of multicast
+ *   is defined as below (https://github.com/p4lang/tutorials/issues/22):
+ *     if (multicast_group != 0)
+ *        multicast_to_group(multicast_group);
+ *     else
+ *        send_to_port(output_port);
+ *
+ *   From this, it follows that if there is a call that sets the
+ *   multicast_group, the packet will be multicast to the group that
+ *   last set the multicast group. Otherwise, the packet will be sent
+ *   to the port set by send_to_port.
+ *
+ * - multiple clone invocations will cause the packet to be cloned to
+ *   the corresponding port. Any drop call in the pipeline will cause
+ *   the packet to drop, and no clone will be created (following the
+ *   analogy with intrinsic metadata bit fields, drop bits are
+ *   processed before clone bits are looked up).
+ *
+ * - resubmit
+ * - recirculate
+ * \TODO: finalize the semantics of calling multiple of the PRE APIs
+ *
+ */
+extern PacketReplicationEngine {
+
+  // PacketReplicationEngine(); /// No constructor. PRE is instantiated
+                                /// by the architecture.
+
+  /// Unicast operation: sends packet to port
+  ///
+  /// Targets may implement this operation by setting the appropriate
+  /// intrinsic metadata or through some other mechanism of
+  /// configuring the PRE.
+  ///
+  /// @param port The output port. If the port is PORT_CPU the packet
+  ///             will be sent to CPU
+  void send_to_port (in PortId_t port);
```

Comment:
Can one query the current value (set by an action in a preceding table)? If not, why not?